### PR TITLE
[RC-1440] Send log section in checks payload

### PR DESCRIPTION
**Pull request by coignetp**
### What does this PR do?

Add a `logs_metadata` section to be able to improve the Fleet Automation view and show the logs integration in this page.
The configuration is sent as a serialized canonical JSON.
To achieve this, it's adding a method in the log interface and putting the log agent as a dependency of the inventorychecks payload.

### Describe how to test/QA your changes

* Configure a log section in an integration (or configure a log integration in any way you like)
* Start the agent
* Checks that the second inventory checks payload contains the `logs_metadata` section

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by julesmcrt — APPROVED** [review]

File: comp/logs/agent/config/integration_config.go [hunk 1]
```diff
@@ -35,59 +35,59 @@ const (
...
 
-	ChannelPath string `mapstructure:"channel_path" json:"channel_path"` // Windows Event
-	Query       string // Windows Event
+	ChannelPath string `mapstructure:"channel_path" json:"channel_path,omitempty"` // Windows Event
+	Query       string `json:"query,omitempty"`                                    // Windows Event
 
 	// used as input only by the Channel tailer.
 	// could have been unidirectional but the tailer could not close it in this case.
-	Channel chan *ChannelMessage
+	Channel chan *ChannelMessage `json:"-"`
```
Comment (gh123man) [inline comp/logs/agent/config/integration_config.go:71]: Do you know what the JSON parser will do with a key of `"-"` (since now multiple fields have that key)? 
This struct is also populated by parsing configs (JSON and YAML). While it's unlikely, I want to avoid any possible breaking changes.

File: comp/logs/agent/config/integration_config.go [hunk 2]
```diff
@@ -35,59 +35,59 @@ const (
 // LogsConfig represents a log source config, which can be for instance
 // a file to tail or a port to listen to.
 type LogsConfig struct {
```
Comment (gh123man) [inline comp/logs/agent/config/integration_config.go:37]: A lot of these settings are set internally and not valid to be set by the user. If we are surfacing this config to the user, do we really want to include all of the internal properties? 

such as (no exhaustive list)
- `Type`
- `Identifier`
- `Query`
- `Image`
- `Label`

I just want to make sure we are not in a position where the user sees all of these config keys and thinks they can change them in the agent.

**Review by gh123man — COMMENTED** [review]
Code LGTM. just a couple of questions

File: comp/metadata/inventorychecks/inventorychecksimpl/inventorychecks.go [hunk 3]
```diff
@@ -17,17 +17,20 @@ import (
...
 	"github.com/DataDog/datadog-agent/pkg/collector"
 	"github.com/DataDog/datadog-agent/pkg/collector/check"
 	checkid "github.com/DataDog/datadog-agent/pkg/collector/check/id"
+	"github.com/DataDog/datadog-agent/pkg/logs/sources"
 	"github.com/DataDog/datadog-agent/pkg/serializer"
 	"github.com/DataDog/datadog-agent/pkg/serializer/marshaler"
 	"github.com/DataDog/datadog-agent/pkg/util/fxutil"
 	"github.com/DataDog/datadog-agent/pkg/util/hostname"
 	"github.com/DataDog/datadog-agent/pkg/util/optional"
+	cjson "github.com/tent/canonical-json-go"
```
Comment (GustavoCaso) [inline comp/metadata/inventorychecks/inventorychecksimpl/inventorychecks.go:33]: I wonder if we need to add a new dependency to our codebase to handle JSON. Could you elaborate why you decided to use an external library, rather the Go JSON library?

Comment (GustavoCaso) [inline comp/metadata/inventorychecks/inventorychecksimpl/inventorychecks.go:33]: I wonder if we need to add a new dependency to our codebase to handle JSON. Could you elaborate on why you decided to use an external library instead of the Go JSON library?

File: comp/metadata/inventorychecks/inventorychecksimpl/inventorychecks.go [hunk 4]
```diff
@@ -197,9 +208,41 @@ func (ic *inventorychecksImpl) getPayload() marshaler.JSONMarshaler {
...
+	}
+
 	return &Payload{
-		Hostname:  ic.hostname,
-		Timestamp: time.Now().UnixNano(),
-		Metadata:  payloadData,
+		Hostname:     ic.hostname,
+		Timestamp:    time.Now().UnixNano(),
+		Metadata:     payloadData,
+		LogsMetadata: logsMetadata,
```
Comment (GustavoCaso) [inline comp/metadata/inventorychecks/inventorychecksimpl/inventorychecks.go:246]: I might be wrong, but I'm not seeing any test checking the new `LogsMetadata` information. If it is true that that we are not checking those in the tests, could we please add some tests 😄

**Review by GustavoCaso — CHANGES_REQUESTED** [review]
I would like to first validate why the use of an external JSON library and add tests if they are missing 😄

File: comp/metadata/inventorychecks/inventorychecksimpl/inventorychecks.go (hunk 3 above)
Comment (coignetp) [reply comp/metadata/inventorychecks/inventorychecksimpl/inventorychecks.go]: I switch it back to the go default library since it's serializing the structure in a consistent order

File: comp/logs/agent/config/integration_config.go (hunk 1 above)
Comment (coignetp) [reply comp/logs/agent/config/integration_config.go]: I did a manual `PublicJSON` export so we're sure not to break anything + addressing your second comment, where we explicitly say the fields we send back to the backend

File: comp/logs/agent/config/integration_config.go (hunk 2 above)
Comment (coignetp) [reply comp/logs/agent/config/integration_config.go]: See https://github.com/DataDog/datadog-agent/pull/21563#discussion_r1430162495

File: comp/metadata/inventorychecks/inventorychecksimpl/inventorychecks.go [hunk 5]
```diff
@@ -197,9 +207,43 @@ func (ic *inventorychecksImpl) getPayload() marshaler.JSONMarshaler {
...
+					continue
+				}
+
+				// Compute log source status
+				logSourceStatus := "pending"
+				if logSource.Status.IsSuccess() {
+					logSourceStatus = "success"
+				} else if logSource.Status.IsError() {
+					logSourceStatus = "error"
+				}
```
Comment (GustavoCaso) [inline comp/metadata/inventorychecks/inventorychecksimpl/inventorychecks.go:230]: The LogStatus has the `Dump()` function. We might be able to add something similar to that struct to get a human-readable string of the status, something like `logSource.Status.String()`

WDYT?

File: comp/logs/agent/mock.go [hunk 6]
```diff
@@ -37,6 +39,18 @@ func newMock(deps dependencies) optional.Option[Mock] {
 	return optional.NewOption[Mock](logsAgent)
 }
 
+// NewMock can be used in other packages using the log agent as a dependency.
+func NewMock(logSources *sources.LogSources) optional.Option[Component] {
```
Comment (gh123man) [inline comp/logs/agent/mock.go:43]: instead of adding a new `NewMock` method (duplicating what the other new mock does), maybe we can re-use what we have in line with how the component framework works. 

I tried to type this as feedback, but it ended up being too big for a comment so I just prototyped it out myself real quick: https://github.com/DataDog/datadog-agent/commit/099a6c1bea62548d9b401ffd9364ab021c1becf0

Let me know what you think or if you have any questions. The idea is to add mock methods to our mock implementation that allows you to mutate it's state, and then use it as a valid component instance. 

(feel free to take my code as is - but it may need a little cleanup)

**Review by gh123man — APPROVED** [review]
AML changes look good to me 👍  Just a suggestion on how to better leverage the component framework and our existing logs agent mock.

File: comp/metadata/inventorychecks/inventorychecksimpl/inventorychecks.go [hunk 7]
```diff
@@ -197,9 +207,43 @@ func (ic *inventorychecksImpl) getPayload() marshaler.JSONMarshaler {
 		}
 	}
 
+	logsMetadata := make(map[string][]metadata)
+	if sources, isSet := ic.sources.Get(); isSet {
+		if sources != nil {
+			for _, logSource := range sources.GetSources() {
+				if _, found := logsMetadata[logSource.Name]; !found {
+					logsMetadata[logSource.Name] = []metadata{}
+				}
```
[the remark this answers is not in the document]
Comment (coignetp) [reply comp/metadata/inventorychecks/inventorychecksimpl/inventorychecks.go:216]: It should not be possible to overwrite the content of `logsMetadata[logSource.Name]`. This condition make sure the key exists if it didn't previously, and after this condition we only append

File: comp/metadata/inventorychecks/inventorychecksimpl/inventorychecks_test.go [hunk 8]
```diff
@@ -156,9 +181,18 @@ func TestGetPayload(t *testing.T) {
 
 	// Check that metadata linked to non-existing check were deleted
 	assert.NotContains(t, "non_running_checkid", ic.data)
+
+	// Check the log sources part of the metadata
+	assert.Len(t, p.LogsMetadata, 1)
+	actualSource, found := p.LogsMetadata["redisdb"]
+	assert.True(t, found)
+	assert.Len(t, actualSource, 1)
+	assert.Equal(t, `{"type":"file","path":"/var/log/redis/redis.log","service":"awesome_cache","source":"redis"}`, actualSource[0]["config"])
```
[the remark this answers is not in the document]
Comment (coignetp) [reply comp/metadata/inventorychecks/inventorychecksimpl/inventorychecks_test.go]: Added the `error` and `status` keys

File: comp/metadata/internal/util/inventory_payload.go [hunk 9]
```diff
@@ -69,7 +69,7 @@ import (
 
 var (
 	defaultMinInterval = 1 * time.Minute
-	defaultMaxInterval = 10 * time.Minute
+	defaultMaxInterval = 1 * time.Minute
```
Comment (GustavoCaso) [inline comp/metadata/internal/util/inventory_payload.go:72]: Is this change intended? What is the difference between `defaultMinInterval` and `defaultMaxInterval`?

**Comment by GustavoCaso** [comment]
Fantastic work 🥇 

As a last ask before approving the PR, I think it is a good practice to add tests for any public methods. We added two new public methods: `LogsConfig.PublicJSON()` and `LogStatus.String()` adding test for those would warranty that there are no unwanted regressions in the future 😄 

After that, the PR would be ready to 🚢 🎉

**Review by GustavoCaso — APPROVED** [review]
Amazing work 🎉 

Thanks you for your patience with my comments @coignetp
